shared/Microsoft.Extensions.TypeNameHelper.Sources/TypeNameHelper.cs

```diff
@@ -36,97 +36,88 @@ public static string GetTypeDisplayName(object item, bool fullName = true)
+        private static void ProcessTypeName(StringBuilder builder, Type type, bool fullName)
         {
-            var totalArgs = args.Length;
-            if (totalArgs >= startIndex + numberOfArgsToAppend)
+            if (type.GetTypeInfo().IsGenericType)
             {
-                sb.Append("<");
-                for (int i = startIndex; i < startIndex + numberOfArgsToAppend; i++)
-                {
-                    ProcessTypeName(args[i], sb, fullName);
-                    if (i + 1 < startIndex + numberOfArgsToAppend)
-                    {
-                        sb.Append(", ");
-                    }
-                }
-                sb.Append(">");
-            }
-        }
-
-        private static void ProcessTypeName(Type t, StringBuilder sb, bool fullName)
-        {
-            if (t.GetTypeInfo().IsGenericType)
-            {
-                ProcessNestedGenericTypes(t, sb, fullName);
+                ProcessGenericType(builder, type, fullName);
                 return;
             }
-            if (_builtInTypeNames.ContainsKey(t))
+            if (_builtInTypeNames.ContainsKey(type))
             {
-                sb.Append(_builtInTypeNames[t]);
+                builder.Append(_builtInTypeNames[type]);
             }
             else
             {
-                sb.Append(fullName ? t.FullName : t.Name);
+                builder.Append(fullName ? type.FullName : type.Name);
             }
         }
 
-        private static void ProcessNestedGenericTypes(Type t, StringBuilder sb, bool fullName)
+        private static void ProcessGenericType(StringBuilder builder, Type type, bool fullName)
         {
-            var genericFullName = t.GetGenericTypeDefinition().FullName;
-            var genericSimpleName = t.GetGenericTypeDefinition().Name;
-            var parts = genericFullName.Split('+');
-            var genericArguments = t.GetTypeInfo().GenericTypeArguments;
-            var index = 0;
-            var totalParts = parts.Length;
-            if (totalParts == 1)
-            {
-                var part = parts[0];
-                var num = part.IndexOf('`');
-                if (num == -1) return;
+            var genericArguments = type.GetGenericArguments();
+            ProcessNestedGenericType(builder, type, genericArguments, genericArguments.Length, fullName);
+        }
 
-                var name = part.Substring(0, num);
-                var numberOfGenericTypeArgs = int.Parse(part.Substring(num + 1));
-                sb.Append(fullName ? name : genericSimpleName.Substring(0, genericSimpleName.IndexOf('`')));
-                AppendGenericArguments(genericArguments, index, numberOfGenericTypeArgs, sb, fullName);
-                return;
+        private static void ProcessNestedGenericType(StringBuilder builder, Type type, Type[] genericArgs, int limit, bool fullName)
```

Comment:
No love for `ArraySegment`? It seems cleaner than passing around a variable called `limit`. I'd prefer the former, but if we need to, perhaps `length` would a better term.

Reply:
I looked at code and mainly you used it for book keeping of where the array ends. An int would also do the same thing. If you think we should really use it then I am fine using ArraySegment. Seemed very low utility to me.

Reply:
I'm a bit disappointed that there are no indexer and `Slice` method for `ArraySegment` in full framework,  but still, it helps to express an idea clearly and after those methods will appear code could be simplified.